// @rc-component/util/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "get", {
  enumerable: true,
  get: function () {
    return _get.default;
  }
});
Object.defineProperty(exports, "merge", {
  enumerable: true,
  get: function () {
    return _set.merge;
  }
});
Object.defineProperty(exports, "mergeWith", {
  enumerable: true,
  get: function () {
    return _set.mergeWith;
  }
});
Object.defineProperty(exports, "noteOnce", {
  enumerable: true,
  get: function () {
    return _warning.noteOnce;
  }
});
Object.defineProperty(exports, "omit", {
  enumerable: true,
  get: function () {
    return _omit.default;
  }
});
Object.defineProperty(exports, "set", {
  enumerable: true,
  get: function () {
    return _set.default;
  }
});
Object.defineProperty(exports, "supportNodeRef", {
  enumerable: true,
  get: function () {
    return _ref.supportNodeRef;
  }
});
Object.defineProperty(exports, "supportRef", {
  enumerable: true,
  get: function () {
    return _ref.supportRef;
  }
});
Object.defineProperty(exports, "toArray", {
  enumerable: true,
  get: function () {
    return _toArray.default;
  }
});
Object.defineProperty(exports, "useComposeRef", {
  enumerable: true,
  get: function () {
    return _ref.useComposeRef;
  }
});
Object.defineProperty(exports, "useControlledState", {
  enumerable: true,
  get: function () {
    return _useControlledState.default;
  }
});
Object.defineProperty(exports, "useEvent", {
  enumerable: true,
  get: function () {
    return _useEvent.default;
  }
});
Object.defineProperty(exports, "useMergedState", {
  enumerable: true,
  get: function () {
    return _useMergedState.default;
  }
});
Object.defineProperty(exports, "warning", {
  enumerable: true,
  get: function () {
    return _warning.default;
  }
});
var _useEvent = _interopRequireDefault(require("./hooks/useEvent"));
var _useMergedState = _interopRequireDefault(require("./hooks/useMergedState"));
var _useControlledState = _interopRequireDefault(require("./hooks/useControlledState"));
var _ref = require("./ref");
var _get = _interopRequireDefault(require("./utils/get"));
var _set = _interopRequireWildcard(require("./utils/set"));
var _warning = _interopRequireWildcard(require("./warning"));
var _omit = _interopRequireDefault(require("./omit"));
var _toArray = _interopRequireDefault(require("./Children/toArray"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
